README.md

```diff
@@ -264,13 +269,92 @@ Command Line Flag | Environment Variable | Default | Description
+[issue tracker](https://github.com/arow-oss/goat-guardian/issues).
+
+- Support other OAuth providers like Google, Github, and Facebook (#1)
+
+- Allow a user to login with *either* email or OAuth and link the accounts.
+
+- Provide a way for Goat Guardian to store user data in a database other than SQLite (#14)
+
+- Allow Goat Guardian settings to be specified in a configuration file (#17)
+
+- Specify "routing-rules" for Goat Guardian, in order to serve static files as
+  well as a web application.  This could also be used to allow Goat Guardian to
+  reverse proxy for multiple web applications, so that users could easy login to
+  multiple related microservices (#18)
+
+- Make Goat Guardian completely programmable, by providing it as a Haskell
+  library (instead of only as a standalone executable).  Eventually we also
+  want to provide access to it as a C library, so that it can be programmed
+  from any programming language, including Python, Ruby, Java, etc.
+
+
+## Related Software
+
+There are a few existing programs that have a similar feature set to Goat
+Guardian: login-with, oauth2_proxy, and kong.  This section compares them to Goat Guardian.
+
+### login-with
+
+[login-with](https://github.com/lipp/login-with) is similar to Goat
+Guardian.  It acts as a "authentication microservice".  It gives an easy way to
+implement authentication in a microservice architecture.
+
+Here are some differences with Goat Guardian:
+
+- Goat Guardian is meant to be used as a reverse-proxy server.  It should sit
+  in front of your normal web application.  It abstracts away the need to worry
+  about authentication in your application.  However, login-with is just
+  another microservice in your microservice-architecture.  Your web application
+  will need to handle the JWT that login-with produces.
+
+- login-with doesn't handle email-based login.  Goat Guardian does.
+
+- login-with can't be run under http (instead of https) for testing.
+
+- The documentation for login-with is lacking.  Manual testing is required to
+  figure out exactly how login-with operates.  If you find the documentation
+  for Goat Guardian lacking in any regard, please open an issue or send a PR.
+
+- login-with can't be used programmatically in Haskell, Python, Ruby, Java, etc.
+
+### oauth2_proxy
+
+[oauth2_proxy](https://github.com/bitly/oauth2_proxy) is very similar to Goat Guardian.  It acts as a reverse proxy in the same way as Goat Guardian.
+
+Here are some differences with Goat Guardian.
+
+- oauth2_proxy doesn't handle email-based registration and login.  Goat Guardian does.
+
+- The documentation for oauth2_proxy is somewhat lacking.  If you find the documentation
+  for Goat Guardian lacking in any regard, please open an issue or send a PR.
+
+- oauth2_proxy only allows you to use one OAuth provider at a time.  Goat
+  Guardian allows you to use any that have been configured.
+
+- login-with can't be used programmatically in Haskell, Python, Ruby, Java, etc.
+
```

Comment:
I guess `oauth2_proxy` only handles authentication process, and just call upstream server with parameter telling oauth2 token to access each OAuth providers.
If the observation above is true, it requires upstream web service developers to add codes to get user information with the token to OAuth providers.

In cases the web service is heavily depends on specific OAuth providers (e.g., a third party twitter client, a service to analyze your activity on GitHub, ...), the `oauth2_proxy` would be nice choice.
(And current version of Goat Guardian cannot be used for such usage because it does not provide a way to pass OAuth tokens to upstream servers)

It seems only for applications using OAuth providers just for identifying logged in users,  Goat Guardian is better than `oauth-proxy` because upstream server only have to check request header to specify which user the logged in user is.

Reply:
The `oauth2_proxy` README is pretty confusing, so I tried to use it to figure out exactly how it works and how it differs from Goat Guardian.

Here is a example of how I ran `oauth2_proxy`:

```sh
$ ./oauth2_proxy \
    -http-address http://localhost:3000 \
    -redirect-url http://localhost:3000/oauth2/callback \
    -upstream 'http://localhost:8000' \
    -cookie-secret 'hellofoobar' \
    -cookie-secure=false \
    -provider github \
    -client-id "t6h197CJY0OdogJ5bUs3" \
    -client-secret "fEMrbXitpU5xNKc0dv3CFbch9YneuELHne1RgH8d" \
    -email-domain '*'
````

This runs `oauth2_proxy` on port 3000 on localhost.  It proxies requests to the upstream server running on localhost:8000.

First, if you go to http://localhost:3000/oauth2/sign_in it presents you with a button you can press to do OAuth with Github.  If you press this button it redirects you to Github to sign in.  Once you sign in, Github redirects you back to http://localhost:3000/oauth2/callback.  This gets your token from Github, and proxies a request to localhost:8000.  The request looks like the following:

```
GET / HTTP/1.1
Host: localhost:3000
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:52.0) Gecko/20100101 Firefox/52.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Encoding: gzip, deflate
Accept-Language: en-US,en;q=0.5
Authorization: Basic c3dvb3NoLm5ld3M6
Cookie: _oauth2_proxy=c3dvb3NoLm5ld3NAZ21haWwuY29t|1535527588|0tpxrgIqg-G2p4rHBFDuQya61A0=
Referer: http://localhost:3000/oauth2/sign_in
Upgrade-Insecure-Requests: 1
X-Forwarded-Email: my.email.address@gmail.com
X-Forwarded-For: 127.0.0.1
X-Forwarded-User: my.github.user
```

You can see that it sets the information from the OAuth token as headers:

```
X-Forwarded-Email: my.email.address@gmail.com
X-Forwarded-For: 127.0.0.1
X-Forwarded-User: my.github.user
```

This could be used in a similar fashion to Goat Guardian.

However, oauth2_proxy has two major flaws:

- `oauth2_proxy` doesn't have any way to reverse proxy endpoints if the user is not logged in.  For instance, if a user who is not logged in tries to access `http://localhost:3000/some/path`, they will be redirected to `http://localhost:3000/oauth2/sign_in`.

    Goat Guardian lets the upstream web application decide which pages to show to the user depending on whether or not they are logged in.

- `oauth2_proxy` can only support one OAuth provider.  It doesn't have its own database of user information.

I will update the README with this information.

Reply:
This has been updated to look like the following in commit 687eef3:

```
[oauth2_proxy](https://github.com/bitly/oauth2_proxy) is very similar to Goat
Guardian.  It acts as a reverse-proxy in the same way as Goat Guardian.

Here are some differences with Goat Guardian.

- oauth2_proxy doesn't handle email-based registration and login.  Goat Guardian does.

- The documentation for oauth2_proxy is somewhat lacking.  If you find the documentation
  for Goat Guardian lacking in any regard, please open an issue or send a PR.

- oauth2_proxy only allows you to use one OAuth provider at a time.  Goat
  Guardian allows you to use any that have been configured.

- oauth2_proxy can't be used programmatically in Haskell, Python, Ruby, Java, etc.

- oauth2_proxy doesn't have any way to reverse-proxy endpoints if the end-user
  hasn't already loggedin.  Goat Guardian lets the upstream application decide
  which pages to show based on whether or not the end-user has already logged
  in.

- oauth2_proxy is mainly used to make internal web applications (for example, a
  self hosted version of [Tiny Tiny RSS](https://tt-rss.org/)) available behind
  OAuth authorization.  Goat Guardian is mainly intended to be used for public
  web applications meant to be used by the general public.

```

Reply:
Nice survey!

> - oauth2_proxy can't be used programmatically in Haskell, Python, Ruby, Java, etc.

This is related to https://github.com/arow-oss/goat-guardian/pull/23#discussion_r213383676